python/dgl/transform.py

```diff
@@ -59,7 +59,7 @@ def pairwise_squared_distance(x):
     return x2s + F.swapaxes(x2s, -1, -2) - 2 * x @ F.swapaxes(x, -1, -2)
 
 #pylint: disable=invalid-name
-def knn_graph(x, k):
+def knn_graph(x, k, use_dgl_impl=False):
```

Comment:
Is nanoflann's KD-tree KNN an *approximate* KNN method?

If so, I agree with Tong's suggestion to rename this (and other similar places') argument.

If not (meaning that nanoflann's implementation is *exact*), I would suggest benchmarking both algorithms and see if C++ is universally more efficient than Python, and simply replace the implementation if this is the case.

Also, did you compare with scikit-learn's [nearest neighbor transformers](https://scikit-learn.org/stable/modules/neighbors.html#nearest-neighbors-transformer)?  If nanoflann is the same as scikit-learn's then I would suggest using that instead.

Reply:
The nanoflann is exact ("No support for approximate NN is provided"). I have tried sklearn's implementation and add experiment results above. It looks like that sklearn's kd-tree performs worse than nanoflann.

Reply:
Based on the above benchmark on `num_points` and `feature_dim`, I think bruteforce and kd-tree algorithms have their own preferred data distribution. 

Besides, when k << n, the current implementation in python that computes an n-by-n distance matrix can be suboptimal, as one can easily avoid the memory consumption by implementing the same algorithm in c++. To claim one is universally more efficient than another, we might need a comprehensive study on various implementations over as many data distributions as possible.

Reply:
By the way, I'm not sure whether to add algorithm option (e.g. "brute", "kd-tree" and "ball-tree") because on GPU we don't use kd-tree and ball-tree.

Reply:
Well we never know, like there are advanced approximation algorithm, they might be added to DGL in the future.

Reply:
Maybe we can add a check here like "if you want to use kd-tree, your data must on CPU".

Reply:
How does KD-tree compare against brute-force (the current implementation) on CPU?  If KD-tree is still better than brute-force then I would say always use KD-tree for CPU and brute-force for GPU, and not provide the algorithm argument.

Reply:
@BarclayII Please check the result above. KD-tree is slower on high-dimensional data, which is common for learned embeddings.

Reply:
KD-tree is better than brute-force provided that the feature dimension `p` is small. When high dimensional data points are used, the KD-tree has poor performance compared to brute-force.

Reply:
I see.  We can either decide which algorithm to use for the users by thresholding, or giving an argument to specify which algorithm to use.

Also, would you mind comparing against `torch.cdist`?  For Minkowski distances the implementation in `pairwise_square_distance` is not the best I would say, especially since we require 1.5.0+ already.

Reply:
Actually the current API only support Euclidean distance (i.e. p=2). I will try the `torch.cdist` to see whether it has better performance typically when p=2.